[examples] Update to support Next.js v9

Janpot opened the pull request:
<!-- Thanks so much for your PR, your contribution is appreciated! ❤️ -->

- [x] I have followed (at least) the [PR section of the contributing guide](https://github.com/mui-org/material-ui/blob/next/CONTRIBUTING.md#submitting-a-pull-request).

* Upcoming release of next has native typescript.
* new typings of `styles` property of the document's initial props is an array of elements. I'm not sure why `flush()` was used here, but might be overlooking something.
* small code style nit to not mutate `ctx`. (I can revert this in case of dislike 🙂)

Draft PR because next@8.1.1 is not released yet (and because maybe some of the other examples can use the same treatment)

oliviertassinari commented on examples/nextjs-next-with-typescript/next.config.js:
```diff
@@ -1,3 +1 @@
-const withTypescript = require('@zeit/next-typescript');
-
-module.exports = withTypescript();
+module.exports = {};
```
Do we need this file?

Janpot replied:
doesn't look like it, I just aligned with the basic next example. Will remove

Janpot replied:
@oliviertassinari Ok, I see now. If I remove this file it tries to pick up the `next.config.js` from the root of the project.

eps1lon commented on examples/nextjs-next-with-typescript/package.json:
```diff
@@ -18,8 +17,6 @@
     "@types/react": "latest",
-    "@types/react-dom": "latest",
-    "@types/styled-jsx": "latest",
-    "typescript": "latest"
```
Don't we need a version of typescript here?

eps1lon commented on examples/nextjs-next/pages/_app.js:
```diff
@@ -9,7 +9,7 @@ class MyApp extends App {
     // Remove the server-side injected CSS.
     const jssStyles = document.querySelector('#jss-server-side');
-    if (jssStyles) {
+    if (jssStyles && jssStyles.parentNode) {
```
What case are we covering here? Seems like the element shouldn't be orphaned in the first place, no?

Janpot replied:
Indeed, but as far as typescript can see, `.parentNode` can be `null`.

eps1lon replied:
Yes for orphaned styles indeed. But this should be considered a bug and not just silently ignored. Since this isn't a typescript file we don't need to concern ourselves with it. 

In a typescript file we would just use `jssStyles.parentNode!.foo` and wait for runtime errors.

oliviertassinari commented:
We can resume the effort https://twitter.com/zeithq/status/1148140410130313218 :)

Janpot commented:
Rebased and fixed the type errors. When running, it doesn't seem like it's serverside rendering for some reason. Don't have any time right now to continue looking into it (on my lunch break).

Janpot commented:
Ok, I'm redoing this one in https://github.com/mui-org/material-ui/pull/16519
